### PR TITLE
Add TTLs to Auth.GetAuthToken, and ExtendAuthToken

### PR DESCRIPTION
**Pull request by msteffen**
Other changes in this PR:
1) The format string for collection indexes is changing. Previously, when a collection element was added to an index, and the index type was e.g. a bool, the key would look like `Collection__index_Field/%!s(bool=true)`. With the migration API that's been added to 1.6.9 and 1.7.0, we can change how index keys are formatted, so I did that in this PR
2) The way TTLs work in collections is also changing. Previously, `transaction.go` had no idea if the values it was caching had TTLs attached to them, meaning there was no way to cache reads/writes of element TTLs; it was the responsibility of `collection.go` to create a new etcd lease when writing a kv with a TTL. Now, that logic has been removed from `collection.go` and read and write caches (`stm.leases` and `stm.newLeases` respectively) have been added to `transaction.go`.

The `newLeases` write cache also preserves the old implementation's desirable property of associating all index kvs with the same lease as the main collection kv.

In other words, if I called `jobs.Put(j)`, the old implementation would do something like:
```
lease := etcdClient.Grant(ttl)
for index := range jobs.Indexes {
   etcdClient.Put(index.Path(), job, WithLease(lease))
}
etcdClient.Put(job, etcdJobInfo, WithLease(lease)
```
The new implementation does something like:
```
Put(key /* = index[0].Path() */, value, ttl) {
  lease := etcdClient.Grant(ttl)
  etcdClient.Put(key, value, WithLease(lease))
}

Put(key /* = index[1].Path() */, value, ttl) {
  lease := newLeases[ttl] // same lease, retrieved from cache
  etcdClient.Put(key, value, WithLease(lease))
}

...
Put(key /* = job */, value /* = etcdJobInfo */, ttl) {
  lease := newLeases[ttl] // still the same lease, retrieved from cache
  etcdClient.Put(key, value, WithLease(lease))
}
```

### CODE REVIEW

File: src/server/auth/server/api_server.go [hunk 1]
```diff
@@ -989,7 +988,92 @@ func (a *apiServer) GetAuthToken(ctx context.Context, req *authclient.GetAuthTok
...
+	if req.TTL > defaultTokenTTLSecs {
+		return nil, fmt.Errorf("can only extend tokens by at most %d seconds", defaultTokenTTLSecs)
+	}
+
+	// The token must already exist. If a token has been revoked, it can't be
+	// extended
+	if _, err := col.NewSTM(ctx, a.etcdClient, func(stm col.STM) error {
+		var (
+			// The collections in which the request token might exist
+			tokens, pipelineTokens = a.tokens.ReadWrite(stm), a.pipelineTokens.ReadWrite(stm)
```
Comment (sjezewski) [inline src/server/auth/server/api_server.go]: This suggests pipeline tokens need to be renewed. I assume pps master handles this? The lines below suggest only `pipelineTokenInfo`s can be issued by admins. So we're issuing `pipelineTokenInfo`s to users of the vault plugin? In the world where non machine users need access to pachyderm via vault, this may need to change.

File: src/server/pkg/collection/collection_test.go [hunk 2]
```diff
@@ -27,8 +30,7 @@ var (
 )
 
 func TestIndex(t *testing.T) {
-	etcdClient, err := getEtcdClient()
-	require.NoError(t, err)
+	etcdClient := getEtcdClient()
```
[the remark this answers is not in the document]
Comment (msteffen) [reply src/server/pkg/collection/collection_test.go]: I started doing this to `getXClient()` methods in tests, after doing it in files that create a client once and then cache the error. When `getXClient()` simply caches an error and returns it in every subsequent invocation, there's no way to recover from the error, so returning it instead of panicking seemed pointless. I subsequently changed `pachyderm_test.go` and `driver/server_test.go`, because those tests never do anything with this error, and making it so that `getXClient()` doesn't return an error makes every test shorter (and means that you can do `getEtcdClient().DoEtcdOperation()` if you only need the client for one thing)

If you disagree, I'm not married to this opinion and I'm happy to change it back, certainly within this file

File: src/server/pkg/collection/transaction.go [hunk 3]
```diff
@@ -281,3 +319,51 @@ func respToValue(key string, resp *v3.GetResponse) (string, error) {
...
+		return ttl, nil
+	}
+
+	// Read kv and lease ID, and cache new TTL
+	getResp := iface.fetch(key) // *** call correct implementation of fetch()
+	if len(getResp.Kvs) == 0 {
+		panic("err not found: " + key)
+		return 0, ErrNotFound{Key: key}
+	}
+	leaseID := v3.LeaseID(getResp.Kvs[0].Lease) // type(kv.Lease) != v3.LeaseID?
```
[the remark this answers is not in the document]
Comment (msteffen) [reply src/server/pkg/collection/transaction.go]: Whoops, leftover note to myself—fixed

File: src/server/pkg/collection/transaction.go [hunk 4]
```diff
@@ -281,3 +319,51 @@ func respToValue(key string, resp *v3.GetResponse) (string, error) {
...
+		return wv.ttl, nil
+	}
+
+	// Read ttl through s.ttlset cache
+	if ttl, ok := s.ttlset[key]; ok {
+		return ttl, nil
+	}
+
+	// Read kv and lease ID, and cache new TTL
+	getResp := iface.fetch(key) // *** call correct implementation of fetch()
```
[the remark this answers is not in the document]
Comment (msteffen) [reply src/server/pkg/collection/transaction.go]: fixed

File: src/server/pkg/collection/transaction.go [hunk 5]
```diff
@@ -281,3 +319,51 @@ func respToValue(key string, resp *v3.GetResponse) (string, error) {
...
+		return 0, ErrNotFound{Key: key}
+	}
+	leaseID := v3.LeaseID(getResp.Kvs[0].Lease) // type(kv.Lease) != v3.LeaseID?
+	if leaseID == 0 {
+		s.ttlset[key] = 0 // 0 is default value, but now 'ok' will be true on check
+		return 0, nil
+	}
+	leaseResp, err := s.client.TimeToLive(s.ctx, leaseID)
+	if err != nil {
+		panic(stmError{err})
```
[the remark this answers is not in the document]
Comment (msteffen) [reply src/server/pkg/collection/transaction.go]: Ah, this is a little bit of a hack, but `Get` and `fetch` do the same thing. Basically, if you look at the implementation of `runSTM`, if one of these calls panics with an `stmError`, then `runSTM` recovers from the panic and passes the error to  the caller.

Basically since we invoke transactions like this:
```
col.NewSTM(ctx, etcdClient, func(stm STM) error {
  ...
  myCol.TT(key) // might break such that we can't commit the txn, but how to communicate that to NewSTM?
})
```
the way this communication happens is `NewSTM` calls this function, which calls `myCol.TTL(key)`. Then `myCol.TTL(key) panics, `NewSTM` calls `recover()`, and then retries the whole transaction without committing.

[the remark this answers is not in the document]
Comment (sjezewski) [reply src/server/pkg/collection/transaction.go]: Got it. Makes sense :+1:

File: src/server/pkg/collection/collection_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (sjezewski) [reply src/server/pkg/collection/collection_test.go]: Ah, I didn't realize we were caching the error. Makes sense
